http-netty/src/main/java/io/micronaut/http/netty/EventLoopSerializer.java

```diff
@@ -0,0 +1,152 @@
+ * limitations under the License.
+ */
+package io.micronaut.http.netty;
+
+import io.micronaut.core.annotation.Internal;
+import io.netty.util.concurrent.EventExecutor;
+import io.netty.util.concurrent.OrderedEventExecutor;
+
+/**
+ * This class forwards reactive operations onto an {@link io.netty.channel.EventLoop} if they are
+ * called from outside that loop. It avoids unnecessary {@link EventExecutor#execute} calls when
+ * the reactive methods are called inside the loop. All the while it ensures the operations remain
+ * serialized (in the same order).
+ * <p>
+ * Should be used like this:
+ * <pre>
+ *     public void onNext(Object item) {
+ *         if (serializer.executeNow(() -> onNext0(item))) {
+ *             onNext0(item);
+ *         }
+ *     }
+ *
+ *     private void onNext0(Object item) {
+ *         ...
+ *     }
+ * </pre>
+ * <p>
+ * This class is <b>not</b> thread-safe: The invariants for calls to {@link #executeNow} are very
+ * strict. In particular:
+ * <ul>
+ *     <li>There must be no concurrent calls to {@link #executeNow}.</li>
+ *     <li>When {@link #executeNow} returns {@code true}, the subsequent execution of the child
+ *     method ({@code onNext0} in the above example) must fully complete before the next
+ *     {@link #executeNow} call. This ensures that there are no concurrent calls to the child
+ *     method.</li>
+ * </ul>
+ * Both of these invariants are guaranteed by the reactive spec, but may not apply to other use
+ * cases.
+ *
+ * @since 4.4.0
+ * @author Jonas Konrad
+ */
+@Internal
+public final class EventLoopSerializer {
+    /**
+     * This adds some extra checks to find bugs.
+     */
+    private static final boolean STRICT_CHECKING = false;
+
+    private final OrderedEventExecutor loop;
+    /**
+     * Generation assigned to the next task.
+     */
+    private int submitGeneration = 0;
+    /**
+     * Generation of the next task that can be executed immediately. All tasks with a lower
+     * generation count have been fully executed already, with one exception: If the last task
+     * was submitted on the event loop, {@link #executeNow} returned true and the caller may not
+     * have fully executed it yet.
+     */
+    private volatile int runGeneration = 0;
+
+    public EventLoopSerializer(OrderedEventExecutor loop) {
+        this.loop = loop;
+    }
+
+    /**
+     * Determine whether the next step can be executed immediately. Iff this method returns
+     * {@code true}, {@code delayTask} will be ignored and the caller should call the target method
+     * manually. Iff this method returns {@code false}, the caller should take no further action as
+     * {@code delayTask} will be run in the future.
+     *
+     * @param delayTask The task to run if it can't be run immediately
+     * @return {@code true} if the caller should instead run the task immediately
+     */
+    public boolean executeNow(Runnable delayTask) {
```

Comment:
I'm not a fan of this signature. Can we execute the runnable?

Reply:
there is a happy path that is only on the event loop where i want to avoid an interface call into the lambda. the goal here is to reduce cost as much as possible for the case where everything is inEventLoop

Reply:
It looks to me like you can implement this as:

```
        if (reentrantLock.isLocked()) {
            // Schedule
        }
        try {
            reentrantLock.lock();
            // Execute
        } finally {
            reentrantLock.unlock();
        }
```

Or with `tryLock`

Reply:
no it really has to run on the event loop so that it does not conflict with other calls that happen on the loop

Reply:
also note that this pattern would not actually solve the serialization. if one task fails to lock, schedules, then a second task succeeds the lock and runs immediately before the first scheduled task runs, we still have incorrect ordering.